src/main/java/org/spongepowered/common/data/manipulator/immutable/entity/ImmutableSpongePassengerData.java

```diff
@@ -0,0 +1,85 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data.manipulator.immutable.entity;
+
+import com.google.common.collect.ComparisonChain;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.MemoryDataContainer;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.immutable.entity.ImmutablePassengerData;
+import org.spongepowered.api.data.manipulator.mutable.entity.PassengerData;
+import org.spongepowered.api.data.value.immutable.ImmutableBoundedValue;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.common.data.ImmutableDataCachingUtil;
+import org.spongepowered.common.data.manipulator.immutable.common.AbstractImmutableData;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongePassengerData;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeBoundedValue;
+
+public class ImmutableSpongePassengerData extends AbstractImmutableData<ImmutablePassengerData, PassengerData> implements ImmutablePassengerData {
+
+    private final Entity vehicle;
+    private final Entity baseVehicle;
+
+    public ImmutableSpongePassengerData(Entity vehicle, Entity baseVehicle) {
+        super(ImmutablePassengerData.class);
+        this.vehicle = vehicle;
+        this.baseVehicle = baseVehicle;
+    }
+
+    @Override
+    public ImmutableBoundedValue<Entity> vehicle() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.VEHICLE, this.vehicle, this.vehicle);
+    }
+
+    @Override
+    public ImmutableBoundedValue<Entity> baseVehicle() {
+        return ImmutableDataCachingUtil.getValue(ImmutableSpongeBoundedValue.class, Keys.BASE_VEHICLE, this.baseVehicle, this.baseVehicle);
+    }
+
+    @Override
+    public ImmutablePassengerData copy() {
+        return new ImmutableSpongePassengerData(this.vehicle, this.baseVehicle);
+    }
+
+    @Override
+    public PassengerData asMutable() {
+        return new SpongePassengerData(this.vehicle, this.baseVehicle);
+    }
+
+    @Override
+    public DataContainer toContainer() {
+        return new MemoryDataContainer()
+                .set(Keys.VEHICLE.getQuery(), this.vehicle)
+                .set(Keys.BASE_VEHICLE.getQuery(), this.baseVehicle);
+    }
+
+    @Override
+    public int compareTo(ImmutablePassengerData o) {
+        return ComparisonChain.start()
+                .compare(o.vehicle().get().getType().getName(), this.vehicle.getType().getName())
+                .compare(o.baseVehicle().get().getType().getName(), this.vehicle.getType().getName())
+                .result();
```

Comment:
compareTo should not return 0 for non equal objects. http://www.javapractices.com/topic/TopicAction.do?Id=10